In [3]:
import pandas as pd

# === Updated Benchmark Results Table with Atom-7B (CPU), Falcon-7B (GPU), Falcon-34B (CPU), and Falcon-40B (CPU) ===
benchmark_results = pd.DataFrame([
    {
        "Task": "Summarization",
        "Model Name": "tiiuae/falcon-7B",
        "Parameters": 6_921_720_704,
        "Model Size (MB)": 26404.27,
        "Latency (s)": 109.560,
        "Tokens": 1220,
        "Tokens/sec": 11.14,
        "Peak RAM (MB)": 420.28,
        "CPU Time (s)": 2925.040,
        "Device": "CPU"
    },
    {
        "Task": "Summarization",
        "Model Name": "tiiuae/falcon-7B (GPU)",
        "Parameters": 6_921_720_704,
        "Model Size (MB)": 13202.13,
        "Latency (s)": 7.008,
        "Tokens": 1220,
        "Tokens/sec": 174.08,
        "Peak RAM (MB)": 693.79,
        "CPU Time (s)": 6.290,
        "Device": "GPU"
    },
    {
        "Task": "Summarization",
        "Model Name": "tiiuae/Falcon-H1-34B-Instruct",
        "Parameters": 33_642_516_224,
        "Model Size (MB)": 128336.01,
        "Latency (s)": 393.751,
        "Tokens": 1222,
        "Tokens/sec": 3.10,
        "Peak RAM (MB)": 1526.61,
        "CPU Time (s)": 9974.790,
        "Device": "CPU"
    },
    {
        "Task": "Summarization",
        "Model Name": "tiiuae/falcon-40b",
        "Parameters": 41_303_293_952,
        "Model Size (MB)": 157559.56,
        "Latency (s)": 1605.262,
        "Tokens": 1220,
        "Tokens/sec": 0.76,
        "Peak RAM (MB)": 802.23,
        "CPU Time (s)": 27967.440,
        "Device": "CPU"
    }
])

# === Display the table ===
benchmark_results


Task                     Model Name   Parameters  Model Size (MB)  \
0  Summarization               tiiuae/falcon-7B   6921720704         26404.27   
1  Summarization         tiiuae/falcon-7B (GPU)   6921720704         13202.13   
2  Summarization  tiiuae/Falcon-H1-34B-Instruct  33642516224        128336.01   
3  Summarization              tiiuae/falcon-40b  41303293952        157559.56   

   Latency (s)  Tokens  Tokens/sec  Peak RAM (MB)  CPU Time (s) Device  
0      109.560    1220       11.14         420.28       2925.04    CPU  
1        7.008    1220      174.08         693.79          6.29    GPU  
2      393.751    1222        3.10        1526.61       9974.79    CPU  
3     1605.262    1220        0.76         802.23      27967.44    CPU

What’s the relationship between memory consumption and latency?

It would seem to me that memory consumption and latency are closely related. According to my benchmarks thus far, models that use higher amounts of RAM tend to see a rise in latency as well, concluding that a model's latency directly correlates with its memory.

torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 2.50 GiB. GPU 0 has a total capacity of 23.49 GiB of which 1.89 GiB is free. Process 1424475 has 6.50 GiB memory in use. Including non-PyTorch memory, this process has 15.09 GiB memory in use. Of the allocated memory 14.79 GiB is allocated by PyTorch, and 14.44 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)


This error was returned when I attempted to run a 34 billion parameter text generation model. This is due to the fact that the model being run has far too many parameters for the 24 GB GPU to handle, 33,642,516,224 to be precise. The GPU doesn't have enough memory to allocate towards this model's processing, thus resulting in a message indicating failure.